packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp

```diff
@@ -411,170 +480,176 @@ preEvaluate(typename TRAITS::PreEvalData d)
-      std::size_t cellLocalId = localCellIds[worksetCellIndex];
-
-      globalIndexer_->getElementGIDs(cellLocalId,GIDs); 
-      blockOffsets[numFieldBlocks] = GIDs.size();
-
-      // caculate the local IDs for this element
-      LIDs.resize(GIDs.size());
-      for(std::size_t i=0;i<GIDs.size();i++) {
-         // used for doing local ID lookups
-         RCP<const MapType> r_map = blockedContainer_->getMapForBlock(GIDs[i].first);
-
-         LIDs[i] = r_map->getLocalElement(GIDs[i].second);
+{
+  using Teuchos::RCP;
+  using Teuchos::rcp_dynamic_cast;
+  using Thyra::VectorBase;
+  using Thyra::ProductVectorBase;
+  using Thyra::BlockedLinearOpBase;
+
+  const auto& localCellIds = this->wda(workset).cell_local_ids_k;
+
+  const int numFieldBlocks = globalIndexer_->getNumFieldBlocks();
+  const RCP<const ContainerType> blockedContainer = blockedContainer_;
+  const RCP<ProductVectorBase<double>> thyraBlockResidual = rcp_dynamic_cast<ProductVectorBase<double>>(blockedContainer_->get_f());
+  const bool haveResidual = Teuchos::nonnull(thyraBlockResidual);
+  const RCP<BlockedLinearOpBase<double>> Jac = rcp_dynamic_cast<BlockedLinearOpBase<double>>(blockedContainer_->get_A(),true);
+
+  // Get the local data for the sub-block crs matrices. First allocate
+  // on host and then deep_copy to device. The sub-blocks are
+  // unmanaged since they are allocated and ref counted separately on
+  // host.
+  using LocalMatrixType = KokkosSparse::CrsMatrix<double,LO,PHX::Device,Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+  typename Kokkos::View<LocalMatrixType**,PHX::Device>::HostMirror
+    hostJacTpetraBlocks("panzer::ScatterResidual_BlockTpetra<Jacobian>::hostJacTpetraBlocks", numFieldBlocks,numFieldBlocks);
+
+  Kokkos::View<int**,PHX::Device> blockExistsInJac =   Kokkos::View<int**,PHX::Device>("blockExistsInJac_",numFieldBlocks,numFieldBlocks);
+  auto hostBlockExistsInJac = Kokkos::create_mirror_view(blockExistsInJac);
+
+  for (int row=0; row < numFieldBlocks; ++row) {
+    for (int col=0; col < numFieldBlocks; ++col) {
+      const auto thyraTpetraOperator = rcp_dynamic_cast<Thyra::TpetraLinearOp<double,LO,GO,NodeT>>(Jac->getNonconstBlock(row,col),false);
+      if (nonnull(thyraTpetraOperator)) {
+
+        // HACK to enforce views in the CrsGraph to be
+        // Unmanaged. Passing in the MemoryTrait<Unmanaged> doesn't
+        // work as the CrsGraph in the CrsMatrix ignores the
+        // MemoryTrait. Need to use the runtime constructor by passing
+        // in points to ensure Unmanaged.  See:
+        // https://github.com/kokkos/kokkos/issues/1581
+
+        // These two lines are the original code we can revert to when #1581 is fixed.
+        // const auto crsMatrix = rcp_dynamic_cast<Tpetra::CrsMatrix<double,LO,GO,NodeT>>(thyraTpetraOperator->getTpetraOperator(),true);
+        // new (&hostJacTpetraBlocks(row,col)) KokkosSparse::CrsMatrix<double,LO,PHX::Device,Kokkos::MemoryTraits<Kokkos::Unmanaged>> (crsMatrix->getLocalMatrix());
```

Comment:
Do we want to create an issue to remind us to switch this at some point?

Reply:
yes